src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -216,6 +218,22 @@ private void SetGlobalErrorIfTrue(bool arg)
             }
         }
 
+        private static void AddCircularStructDiagnostics(NamespaceOrTypeSymbol symbol)
+        {
+            if (symbol is SourceMemberContainerTypeSymbol sourceMemberContainerTypeSymbol)
```

Comment:
>symbol is SourceMemberContainerTypeSymbol sourceMemberContainerTypeSymbol [](http://example.com/codeflow?start=16&length=73)

It looks like we are ignoring ```filterOpt```

Reply:
I'll add the PassesFilter check here.

Pending a unit test that gets affected.

Reply:
Added the check. Pending a unit test

Reply:
@AlekseyTs I couldn't figure a unit test. I tried with `InStruct_Cycle` which uses a filter to get diagnostics of a specific tree (let's say S1). When we reach `AddCircularStructDiagnostics`, it turned out that `KnownCircularStruct` for S2 was already accessed earlier.

IT's `DefiniteAssignmentPass` that is accessing `KnownCircularStruct` for S2 when it's analyzing the property in S1.

Relevant part of the stack trace:

![image](https://user-images.githubusercontent.com/31348972/166093425-95180b2a-db46-4c91-bd8a-f2080ce92e47.png)

I don't know if `filterOpt` check can have any effect in any other scenario, so better to keep it even if no test coverage for that?

I'm not sure if that's related to this?

https://github.com/dotnet/roslyn/blob/55bc2b7a1d957238a6cb75682850f27f4584733f/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3053-L3056